curvlinops/jacobian.py

```diff
@@ -61,110 +63,55 @@ def __init__(
-
-        rtol, atol = 5e-5, 1e-6
-
-        def check_X_y(X1, X2, y1, y2):
-            if not allclose(X1, X2) or not allclose(y1, y2):
-                self.print_nonclose(X1, X2, rtol=rtol, atol=atol)
-                self.print_nonclose(y1, y2, rtol=rtol, atol=atol)
-                raise RuntimeError("Non-deterministic data loading detected.")
-
-        with no_grad():
-            for (X1, y1), (X2, y2) in zip(
-                self._loop_over_data(desc="_check_deterministic_data_pred"),
-                self._loop_over_data(desc="_check_deterministic_data_pred2"),
-            ):
-                pred1, y1 = self._model_func(X1).cpu().numpy(), y1.cpu().numpy()
-                pred2, y2 = self._model_func(X2).cpu().numpy(), y2.cpu().numpy()
-
-                if isinstance(X1, Tensor) or isinstance(X2, Tensor):
-                    X1, X2 = X1.cpu().numpy(), X2.cpu().numpy()
-                    check_X_y(X1, X2, y1, y2)
-                else:  # X is a MutableMapping
-                    for k in X1.keys():
-                        v1, v2 = X1[k], X2[k]
-
-                        if isinstance(v1, Tensor) or isinstance(v2, Tensor):
-                            X1, X2 = v1.cpu().numpy(), v2.cpu().numpy()
-
-                        check_X_y(X1, X2, y1, y2)
+        x = next(iter(self._data))[0]
+        if isinstance(x, Tensor):
+            x = x.to(self._infer_device())
 
-                if not allclose(pred1, pred2):
-                    self.print_nonclose(pred1, pred2, rtol=rtol, atol=atol)
-                    raise RuntimeError("Non-deterministic model detected.")
+        return [(self._N_data,) + self._model_func(x).shape[1:]]
 
-    def _matmat(self, M: ndarray) -> ndarray:
-        """Apply the Jacobian to a matrix.
+    def _matmat(self, M: List[Tensor]) -> List[Tensor]:
+        """Apply the Jacobian to a matrix in tensor list format.
 
         Args:
-            M: Matrix for multiplication. Has shape ``[D, K]``.
+            M: Matrix for multiplication in tensor list format.
 
         Returns:
-            Matrix-multiplication result ``J @ M``. Has shape ``[N * C, K]``.
+            Matrix-multiplication result ``J @ M`` in tensor list format.
         """
-        num_vecs = M.shape[1]
-        M_list = self._preprocess(M)
-
-        result_list = []
+        (num_vecs,) = {m.shape[-1] for m in M}
+        JM = []
 
         for X, _ in self._loop_over_data(desc="_matmat"):
             output = self._model_func(X)
 
             # multiply the mini-batch Jacobian onto all vectors
-            col = []
-            for n in range(num_vecs):
-                (col_n,) = jvp(output, self._params, [M[n] for M in M_list])
-                col.append(col_n.flatten(start_dim=1))
-            # combine columns into a single tensor and append
-            result_list.append(stack(col))
+            JM_col = [
+                jvp(output, self._params, [m[..., n] for m in M])[0]
+                for n in range(num_vecs)
```

Comment:
Yes, this would be the best way to remove the `for`-loop. To do that, I would have to refactor BackPACK's `jvp` function. Also, this exposes some new hyper-parameters, for instance we want to chunk the `vmap` if we would run out of memory doing the `jvp` for all vectors in one go.

Reply:
For scalability (in terms of memory consumption), the `for` loop is ideal.

Reply:
Is there a reason not to use `torch.func.jvp`? But I was also thinking about the potential memory issues...